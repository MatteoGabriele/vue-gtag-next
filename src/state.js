import { ref, reactive, computed, toRefs } from "vue";

const state = reactive({
  property: null,
  isEnabled: true,
  globalObjectName: "gtag",
  globalDataLayerName: "dataLayer",
  domain: "https://www.googletagmanager.com",
  customResource: null,
  useDebugger: false,
});

export const routeState = reactive({
  template: null,
  appName: null,
  useScreenview: false,
  skipSamePath: true,
});

export const useState = () => toRefs(state);

export const useRouteState = () => toRefs(routeState);

export const isBootstrapped = ref(false);

export const isReady = ref(false);

export const isTrackRouterEnabled = ref(false);

export const defaultProperty = computed(() => {
  const { property } = useState();

  if (!property.value) {
    return;
  }

  if (Array.isArray(property.value)) {
    return property.value.find((p) => p.default === true) || property.value[0];
  }

  return property.value;
});

export const hasId = computed(() => {
  const { property } = useState();
  return property.value && property.value.id !== null;
});

export const allProperties = computed(() => {
  const { property } = useState();

  if (Array.isArray(property.value)) {
    return property.value;
  }

  return [property.value];
});

export const isTracking = computed(() => {
  const { isEnabled } = useState();
  const property = defaultProperty.value;

  return property && property.id && isEnabled.value;
});

export default state;
