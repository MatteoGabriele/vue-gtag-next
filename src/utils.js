export const loadScript = (source, domain) => {
  return new Promise((resolve, reject) => {
    const head = document.head || document.getElementsByTagName("head")[0];
    const script = document.createElement("script");

    script.async = true;
    script.src = source;
    script.charset = "utf-8";

    if (domain) {
      const link = document.createElement("link");

      link.href = domain;
      link.rel = "preconnect";

      head.appendChild(link);
    }

    head.appendChild(script);

    script.onload = resolve;
    script.onerror = reject;
  });
};

export const merge = (obj = {}, newObj = {}) => {
  Object.keys(newObj).forEach((key) => {
    obj[key] = newObj[key];
  });
};
