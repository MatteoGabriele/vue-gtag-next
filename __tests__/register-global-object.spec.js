import state from "@/state";
import { merge } from "@/utils";
import registerGlobalObject from "@/register-global-object";
import mockdate from "mockdate";

mockdate.set("2020-01-01T01:01:01Z");

const defaultState = { ...state };

describe("register-global-object", () => {
  beforeEach(() => {
    merge(state, defaultState);
  });

  afterEach(() => {
    delete global[state.globalObjectName];
    delete global[state.globalDataLayerName];

    jest.restoreAllMocks();
    jest.clearAllMocks();
  });

  it("should register gtag global object", () => {
    registerGlobalObject();
    expect(window.gtag).toBeDefined();
  });

  it("should register gtag global object under another name", () => {
    merge(state, { globalObjectName: "foo" });
    registerGlobalObject();

    expect(window.foo).toBeDefined();
    expect(window.gtag).not.toBeDefined();
  });

  it("should push arguments inside the dataLayer", () => {
    registerGlobalObject();

    window.gtag("foo", "bar");

    expect(window.dataLayer).toMatchSnapshot();
  });
});
