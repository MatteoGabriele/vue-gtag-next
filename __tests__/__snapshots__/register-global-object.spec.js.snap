// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register-global-object should push arguments inside the dataLayer 1`] = `
Array [
  Arguments [
    "js",
    2020-01-01T01:01:01.000Z,
  ],
  Arguments [
    "foo",
    "bar",
  ],
]
`;
